File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/VB/WebProject1/WebProject1/Services/TempMailTaskStore.vb
Public Class TempMailTaskStore
    Private Shared _tasks As Dictionary(Of String, MailTask) = Nothing
    Private Shared _lock As Object = Nothing

    Public Shared Sub Init()
        _lock = New Object()
        _tasks = New Dictionary(Of String, MailTask)()
    End Sub

    Public Shared Sub PutTask(ByVal mailTask As MailTask)
        SyncLock _lock

            If Not _tasks.ContainsKey(mailTask.TaskId) Then
                _tasks.Add(mailTask.TaskId, mailTask)
            End If
        End SyncLock
    End Sub

    Public Shared Function GetTask(ByVal taskId As String) As MailTask
        Dim task As MailTask = Nothing

        If String.IsNullOrEmpty(taskId) Then
            Return task
        End If

        SyncLock _lock

            If _tasks.ContainsKey(taskId) Then
                task = _tasks(taskId)
                _tasks.Remove(taskId)
            End If
        End SyncLock

        Return task
    End Function
End Class

File: Pacman-Game-OOP-master 2/EASendMail/Samples_ASPNetMvc/C#/WebProject1/WebProject1/Services/TempMailTaskStore.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using WebProject1.Models;

namespace WebProject1.Services
{
    public class TempMailTaskStore
    {
        private static Dictionary<string, MailTask> _tasks = null;
        private static object _lock = null;

        public static void Init()
        {
            _lock = new object();
            _tasks = new Dictionary<string, MailTask>();
        }

        public static void PutTask(MailTask mailTask)
        {
            lock (_lock)
            {
                if (!_tasks.ContainsKey(mailTask.TaskId))
                {
                    _tasks.Add(mailTask.TaskId, mailTask);
                }
            }
        }

        public static MailTask GetTask(string taskId)
        {
            MailTask task = null;
            if (string.IsNullOrEmpty(taskId))
            {
                return task;
            }

            lock (_lock)
            {
                if (_tasks.ContainsKey(taskId))
                {
                    task = _tasks[taskId];
                    _tasks.Remove(taskId);
                }
            }

            return task;
        }
    }
}